### ISSUE TITLE
Podman cannot use CUDA under rootless container.

### ISSUE THREAD

**Issue by Sumi-Sumi** [containers/podman#9926]
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Podman container can see GPUs (using `nvidia-smi`) but cannot use CUDA.

**Steps to reproduce the issue:**
I used rootless podman.

1. Pull PyTorch image from docker.io (`podman pull docker.io/pytorch/pytorch:1.6.0-cuda10.1-cudnn7-devel`)

2. Set up nvidia-contianer-toolkit and podman profile. ( I watched [this(Red Hat page).](https://www.redhat.com/ja/blog/how-use-gpus-containers-bare-metal-rhel-8))
I didn't use SELinux so I skip step relating to SELinux.

3.  Run container with `podman run --security-opt label=type:nvidia_container_t --hook-dir=/usr/share/container/oci/hooks.d/ docker.io/pytorch/pytorch:1.6.0-cuda10.1-cudnn7-devel  /bin/bash` or `podman run -it --rm docker.io/pytorch/pytorch:1.6.0-cuda10.1-cudnn7-devel  /bin/bash` and `python > import torch > torch.cuda.is_available()`

**Describe the results you received:**

```
> python
> import torch
> torch.cuda.is_available()
>> False
```
I tried one more after I did the process on host. (My host returned true.)
```
> python
> import torch
> torch.cuda.is_available()
>> True
```

**Describe the results you expected:**
I expect the result before I do the process on host.
```
> python
> import torch
> torch.cuda.is_available()
>> True
```

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
> podman version   
Version:      3.1.0
API Version:  3.1.0
Go Version:   go1.16.2
Git Commit:   9f09fb62cba8f44c18eda84db3e72aab3ae76046-dirty
Built:        Wed Mar 31 23:46:48 2021
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
> podman info --debug        
host:
  arch: amd64
  buildahVersion: 1.20.0
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: /usr/bin/conmon is owned by conmon 1:2.0.27-1
    path: /usr/bin/conmon
    version: 'conmon version 2.0.27, commit: 65fad4bfcb250df0435ea668017e643e7f462155'
  cpus: 24
  distribution:
    distribution: arch
    version: unknown
  eventLogger: journald
  hostname: **
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 300000
      size: 200000
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 300000
      size: 200000
  kernel: 5.11.11-arch1-1
  linkmode: dynamic
  memFree: 26095562752
  memTotal: 33584173056
  ociRuntime:
    name: runc
    package: /usr/bin/runc is owned by runc 1.0.0rc93-2
    path: /usr/bin/runc
    version: |-
      runc version 1.0.0-rc93
      commit: 12644e614e25b05da6fd08a38ffa0cfe1903fdec
      spec: 1.0.2-dev
      go: go1.16.2
      libseccomp: 2.5.1
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: true
    seccompEnabled: true
    selinuxEnabled: false
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: /usr/bin/slirp4netns is owned by slirp4netns 1.1.9-1
    version: |-
      slirp4netns version 1.1.9
      commit: 4e37ea557562e0d7a64dc636eff156f64927335e
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.1
  swapFree: 25165819904
  swapTotal: 25165819904
  uptime: 1h 8m 35.99s (Approximately 0.04 days)
registries: {}
store:
  configFile: /home/sumi/.config/containers/storage.conf
  containerStore:
    number: 2
    paused: 0
    running: 0
    stopped: 2
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: /usr/bin/fuse-overlayfs is owned by fuse-overlayfs 1.5.0-1
      Version: |-
        fusermount3 version: 3.10.2
        fuse-overlayfs: version 1.5
        FUSE library version 3.10.2
        using FUSE kernel interface version 7.31
  graphRoot: /home/sumi/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 5
  runRoot: /run/user/1000/containers
  volumePath: /home/sumi/.local/share/containers/storage/volumes
version:
  APIVersion: 3.1.0
  Built: 1617202008
  BuiltTime: Wed Mar 31 23:46:48 2021
  GitCommit: 9f09fb62cba8f44c18eda84db3e72aab3ae76046-dirty
  GoVersion: go1.16.2
  OsArch: linux/amd64
  Version: 3.1.0
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
> pacman -Qi podman                           
Name            : podman
Version         : 3.1.0-1
Description     : Tool and library for running OCI-based containers in pods
Architecture    : x86_64
URL             : https://github.com/containers/libpod
Licenses        : Apache
Groups          : None
Provides        : None
Depends On      : cni-plugins  conmon  containers-common  device-mapper  iptables  libseccomp  runc  slirp4netns  libsystemd  fuse-overlayfs  libgpgme.so=11-64
Optional Deps   : podman-docker: for Docker-compatible CLI
                  btrfs-progs: support btrfs backend devices [installed]
                  catatonit: --init flag support
                  crun: support for unified cgroupsv2
Required By     : None
Optional For    : None
Conflicts With  : None
Replaces        : None
Installed Size  : 72.86 MiB
Packager        : Morten Linderud <foxboron@archlinux.org>
Build Date      : Wed 31 Mar 2021 11:46:48 PM JST
Install Date    : Fri 02 Apr 2021 02:28:13 AM JST
Install Reason  : Explicitly installed
Install Script  : No
Validated By    : Signature
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
OS: Arch Linux x86_64
Kernel: 5.11.11-arch1-1
CPU: AMD Ryzen 9 3900X
GPU: NVIDIA GeForce RTX 2060 SUPER
GPU: NVIDIA GeForce RTX 2070 SUPER

**Comment by mheon** [comment]
Running as root, or rootless?

**Comment by Sumi-Sumi** [comment]
I runed as rootless.
I made the necessary settings for rootless.
`nvidia-smi` returned normal reply without error but I could not use CUDA for the first launch.
I will change title.

**Comment by zvonkok** [comment]
@Sumi-Sumi If you do not use SELinux then do not use a SELinux label aka context `nvidia_container_t` 
`--security-opt label=type:nvidia_container_t` this is SELinux related. 

Try to run without the `security-opt` if it still fails try privileged if this succeeds we may have wrong privs on the device files.  You can also try to run the container with `/bin/bash`  and to some `ls -Zla /dev/nvidia*`

**Comment by Sumi-Sumi** [comment]
I tried running container with `--privileged` under root user.
The result is good.
However, I failed under normal user. (I tried both adding `--privileged` and no adding it.)
I put the result of `ls -Zla /dev/nvidia*`.
The container returned "nobody" so I think I'm lacking authority under rootless.
Is there a way to solve this?

* Under normal user
```
crw-rw-rw- 1 nobody nogroup ? 195,   0 Apr  7 06:52 /dev/nvidia0
crw-rw-rw- 1 nobody nogroup ? 195, 255 Apr  7 06:52 /dev/nvidiactl
```

* Under root user
```
crw-rw-rw- 1 root root ? 195,   0 Apr  7 06:52 /dev/nvidia0
crw-rw-rw- 1 root root ? 195, 255 Apr  7 06:52 /dev/nvidiactl
```

**Comment by mheon** [comment]
The normal user doesn't seem to have full permissions on the nvidia devices (specifically lacking execute) - that seems like a potential culprit?

**Comment by Sumi-Sumi** [comment]
I observed a difference in behavior between the root user and the non-root user, and it seems that `nvidia-uvm` is not being loaded.
https://github.com/NVIDIA/nvidia-container-runtime/issues/85  shows how to run it routelessly by setting up udev rules.
We believe that this can be used for this problem as well.
However, I don't know how to write udev rules related pci-gpu device.

**Comment by Sumi-Sumi** [comment]
I solved this problem.
When launching a container from a non-root user, CUDA cannot be used from within the container because the device node is not created.
I did some more research and succeeded in creating udev.
In nvidia drive ver, 460.67, it was necessary to use `nvidia-modprobe` instead of `create-uvm-dev-node` as shown in https://github.com/NVIDIA/nvidia-container-runtime/issues/85.
Here are the udev rules I used.

```
# Load and unload nvidia-uvm module
ACTION=="add", DEVPATH=="/bus/pci/drivers/nvidia", RUN+="/sbin/modprobe nvidia-uvm"
ACTION=="remove", DEVPATH=="/bus/pci/drivers/nvidia", RUN+="/sbin/modprobe -r nvidia-uvm"

# This will create the device nvidia device nodes
ACTION=="add", DEVPATH=="/bus/pci/drivers/nvidia", RUN+="/usr/bin/nvidia-smi"

# Create the device node for the nvidia-uvm module
ACTION=="add", DEVPATH=="/module/nvidia_uvm", SUBSYSTEM=="module", RUN+="/usr/bin/nvidia-modprobe -u -c=0"
```

**Comment by rhatdan** [comment]
You would be best off asking this somewhere else, I believe, unless @nalind or @mtrmac have any idea on udev rules?

**Comment by rhatdan** [comment]
Since no follow up, closing, Reopen if you believe there is something podman can do.
